js/src/admin/components/UserListPage.tsx

```diff
@@ -0,0 +1,383 @@
+            >
+              {user.username()}
+            </a>
+          );
+        },
+      },
+      90
+    );
+
+    columns.add(
+      'joinDate',
+      {
+        name: app.translator.trans('core.admin.users.grid.default_columns.join_time'),
+        content: (user: User) => (
+          <span class="UserList-joinDate" title={user.joinTime()}>
+            {dayjs(user.joinTime()).format('LLL')}
+          </span>
+        ),
+      },
+      80
+    );
+
+    columns.add(
+      'groupBadges',
+      {
+        name: app.translator.trans('core.admin.users.grid.default_columns.group_badges'),
+        content: (user: User) => {
+          const badges = user.badges().toArray();
+
+          if (badges.length) {
+            return <ul className="DiscussionHero-badges badges">{listItems(badges)}</ul>;
+          } else {
+            return app.translator.trans('core.admin.users.grid.default_columns.group_badges_none');
+          }
+        },
+      },
+      70
+    );
+
+    columns.add(
+      'emailAddress',
+      {
+        name: app.translator.trans('core.admin.users.grid.default_columns.email'),
+        content: (user: User) => {
+          function toggleEmailVisibility() {
+            // Get needed jQuery element refs
+            const emailContainer = $(`[data-column-name=emailAddress][data-user-id=${user.id()}] .UserList-email`);
+            const emailAddress = emailContainer.find('.UserList-emailAddress');
+            const emailToggleButton = emailContainer.find('.UserList-emailIconBtn');
+            const emailToggleButtonIcon = emailToggleButton.find('.icon');
+
+            const emailShown = emailContainer.attr('data-email-shown') === 'true';
+
+            if (emailShown) {
+              //! Email currently shown, switching to hidden
+
+              // Update tooltip
+              emailToggleButton.attr('title', extractText(app.translator.trans('core.admin.users.grid.default_columns.email_visibility_show')));
+
+              // Replace real email with placeholder email
+              emailAddress.text(app.translator.trans('core.admin.users.grid.default_columns.email_hidden'));
+              emailAddress.attr('aria-hidden', 'true');
+
+              // Change button icons
+              emailToggleButtonIcon.removeClass('fa-eye');
+              emailToggleButtonIcon.addClass('fa-eye-slash');
+            } else {
+              //! Email currently hidden, switching to shown
+
+              // Update tooltip
+              emailToggleButton.attr('title', extractText(app.translator.trans('core.admin.users.grid.default_columns.email_visibility_hide')));
+
+              // Replace placeholder email with real email
+              emailAddress.text(user.email());
+              emailAddress.removeAttr('aria-hidden');
+
+              // Change button icons
+              emailToggleButtonIcon.addClass('fa-eye');
+              emailToggleButtonIcon.removeClass('fa-eye-slash');
+            }
```

Comment:
The code is auto descriptive and well organised, I don't think these comments are necessary.